src/core/room.ts

```diff
@@ -129,53 +138,176 @@ export class DefaultRoom implements Room {
+        roomId,
+        options,
+        realtime,
+        realtime.auth.clientId,
+        logger,
+        initFeaturesAfter,
+      );
       features.push(this._presence);
     }
 
     if (options.typing) {
       this._logger.debug('enabling typing on room', { roomId });
-      this._typing = new DefaultTyping(roomId, options.typing, realtime, realtime.auth.clientId, logger);
+      this._typing = new DefaultTyping(
+        roomId,
+        options.typing,
+        realtime,
+        realtime.auth.clientId,
+        logger,
+        initFeaturesAfter,
+      );
       features.push(this._typing);
     }
 
     if (options.reactions) {
       this._logger.debug('enabling reactions on room', { roomId });
-      this._reactions = new DefaultRoomReactions(roomId, realtime, realtime.auth.clientId, logger);
+      this._reactions = new DefaultRoomReactions(roomId, realtime, realtime.auth.clientId, logger, initFeaturesAfter);
       features.push(this._reactions);
     }
 
     if (options.occupancy) {
       this._logger.debug('enabling occupancy on room', { roomId });
-      this._occupancy = new DefaultOccupancy(roomId, realtime, this._chatApi, logger);
+      this._occupancy = new DefaultOccupancy(roomId, realtime, this._chatApi, logger, initFeaturesAfter);
       features.push(this._occupancy);
     }
 
-    // Setup lifecycle manager - reverse the features so messages always comes in last
-    this._lifecycleManager = new RoomLifecycleManager(this._status, features.toReversed(), logger, 5000);
-
-    // Setup a finalization function to clean up resources
-    let finalized = false;
-    this._finalizer = async () => {
-      // Cycle the channels in the feature and release them from the realtime client
-      if (finalized) {
-        this._logger.debug('Room.finalizer(); already finalized');
-        return;
+    // Wait for features to finish initializing and then finish initializing
+    // the room. Set _asyncOpsAfter to the promise that waits for the room to
+    // finish initializing. This promise is awaited before performing any async
+    // operations at room level (attach and detach).
+    this._asyncOpsAfter = initFeaturesAfter.then(() => {
+      // Features have now started initializing so we can no longer stop the
+      // initialization process. Features haven't yet finished initializing,
+      // so if release() is called we first need to wait for initialization to
+      // finish before releasing. We use a promise to wait for the correct
+      // release function.
+
+      let setFinalizerFunc: (f: () => Promise<void>) => void;
+      const finalizerFuncPromise = new Promise<() => Promise<void>>((resolve) => {
+        setFinalizerFunc = resolve;
+      });
+      this._finalizer = () => {
+        return finalizerFuncPromise.then((f) => {
+          return f();
+        });
+      };
+
+      // Setup all contributors with resolved channels
+      interface ContributorWithChannel {
```

Comment:
Another nitpick, should we move this interface outside the constructor and document it? Personally I think its better to define interfaces at the top level, or do you think it's best here?

Reply:
removed it entirely